csrc/preseg_passes/allocation_order_inference.cpp

```diff
@@ -15,442 +16,279 @@ namespace nvfuser::preseg_passes {
-    const std::vector<Val*>& candidates) {
-  TensorView* src = nullptr;
-  size_t non_bc_high_water_mark = 0;
-
-  // helper utils to count the number of non broadcast / non reduction
-  // iterdomain
-  auto countLoopIterDomains = [](const TensorView* tv) -> size_t {
-    return std::count_if(
-        tv->getMaybeRFactorDomain().begin(),
-        tv->getMaybeRFactorDomain().end(),
-        [&](auto ptr_id) {
-          return !ptr_id->isBroadcast() && !ptr_id->isReduction();
-        });
-  };
-
-  for (auto* val : candidates) {
-    auto* tv = dynamic_cast<TensorView*>(val);
-    // skip non TensorView entry
-    if (tv == nullptr) {
+  // removing mapped ids and reduction ids to create unmapped_ids_vec.
+  // This means for the rest of ids in target_rfactor_domain that's not in
+  // mapped_id_set, they are either 1. a reduction domain, or; 2. in
+  // [unmapped_ids_vec.begin(), unmapped_ids_vec_end) This ensures that
+  // sharp-edges 1's loop would reconstruct a permutation of the
+  // target_rfactor_domain, hence a valid allocation domain for target.
+  std::vector<IterDomain*> unmapped_ids_vec = target_rfactor_domain;
+  auto unmapped_ids_vec_end = std::remove_if(
+      unmapped_ids_vec.begin(),
+      unmapped_ids_vec.end(),
+      [&mapped_id_set](IterDomain* it) {
+        return mapped_id_set.count(it) != 0 || it->isReduction();
+      });
+
+  auto mapped_id_iter = mapped_id_vec.begin();
+  auto unmapped_id_iter = unmapped_ids_vec.begin();
+  // initialize new target allocation domain with nullptr
+  std::vector<IterDomain*> target_alloc_domain(
+      target_rfactor_domain.size(), nullptr);
+  for (auto i : c10::irange(target_rfactor_domain.size())) {
+    // sharp-edges 1
+    // preserves non-mapped reduction id in its original position
+    if (target_rfactor_domain[i]->isReduction() &&
+        mapped_id_set.count(target_rfactor_domain[i]) == 0) {
+      target_alloc_domain[i] = target_rfactor_domain[i];
       continue;
     }
-
-    auto iter = alloc_order_map_.find(tv);
-    // stopping propagation when we encounter an entry that does not have an
-    // allocation order. See NOTE: [Allocation Order Inference]
-    if (iter == alloc_order_map_.end()) {
-      return nullptr;
+    // push unmapped ids to outer dimension until it's fully consumed
+    if (unmapped_id_iter != unmapped_ids_vec_end) {
+      target_alloc_domain[i] = *unmapped_id_iter++;
+    } else {
+      // push mapped ids to inner dimension
+      target_alloc_domain[i] = *mapped_id_iter++;
     }
-
-    // skip entry that has an empty allocation order
-    if (iter->second.empty()) {
-      // We still want to ensure that we propagate empty allocation order if
-      // there's no candidate with a non-empty allocation order
-      if (src == nullptr) {
-        src = tv;
+  }
+#else
+  // mapping id between ref's allocation domain to target's rfactor domain
+  for (auto* ref_id : ref_alloc_domain) {
+    for (auto* id : target_rfactor_domain) {
```

Comment:
Instead of using two nested for loop, would it be better to use a ValGroup->IterDomain map mapping the group of each ID in `target_rfactor_domain` to the ID itself?